File: internal/client/envelope.go
package client

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/zatiti/zatiti/internal/contract"
)

// operationIDMax bounds an operation path segment defensively; the frozen
// catalog's dot-separated IDs stay far below it.
const operationIDMax = 200

// submissionKeyMax is the frozen submission key bound: 1..128 printable
// ASCII characters.
const submissionKeyMax = 128

// marshalRequest validates the operation identifier and request envelope and
// renders the exact request body bytes. The input is validated with zatiti
// wire strictness (valid UTF-8, no duplicate keys, no trailing data,
// integers within int64) so caller-side mistakes fail before anything is
// sent. A nil input marshals as JSON null and the controller's schema
// decides whether that is acceptable.
func marshalRequest(operation string, req contract.Request) ([]byte, error) {
	if err := validateOperationID(operation); err != nil {
		return nil, err
	}
	switch req.Schema {
	case "":
		req.Schema = contract.SchemaRequest
	case contract.SchemaRequest:
	default:
		return nil, fmt.Errorf("%w: request schema %q, want %q", ErrInvalidRequest, req.Schema, contract.SchemaRequest)
	}
	if req.SubmissionKey != "" {
		if err := validateSubmissionKey(req.SubmissionKey); err != nil {
			return nil, err
		}
	}
	if len(req.Input) > 0 {
		if err := contract.DecodeStrict(req.Input, new(json.RawMessage)); err != nil {
			return nil, fmt.Errorf("%w: input is not strict JSON: %v", ErrInvalidRequest, err)
		}
	}
	body, err := json.Marshal(req)
	if err != nil {
		return nil, fmt.Errorf("%w: request cannot be encoded: %v", ErrInvalidRequest, err)
	}
	return body, nil
}

// validateOperationID accepts the frozen catalog shape: non-empty
// dot-separated segments of [a-z0-9_], no empty segment, no leading or
// trailing dot. The identifier is used verbatim as the single path segment,
// so anything else is refused rather than escaped.
func validateOperationID(operation string) error {
	if operation == "" {
		return fmt.Errorf("%w: operation is required", ErrInvalidRequest)
	}
	if len(operation) > operationIDMax {
		return fmt.Errorf("%w: operation identifier exceeds %d bytes", ErrInvalidRequest, operationIDMax)
	}
	for i := 0; i < len(operation); i++ {
		c := operation[i]
		if c == '.' || c == '_' || ('a' <= c && c <= 'z') || ('0' <= c && c <= '9') {
			continue
		}
		return fmt.Errorf("%w: operation %q contains a byte outside [a-z0-9_.]", ErrInvalidRequest, operation)
	}
	for _, segment := range strings.Split(operation, ".") {
		if segment == "" {
			return fmt.Errorf("%w: operation %q has an empty segment", ErrInvalidRequest, operation)
		}
	}
	return nil
}

// validateSubmissionKey enforces the frozen bound: 1..128 printable ASCII
// characters (0x20..0x7E).
func validateSubmissionKey(key string) error {
	if key == "" {
		return fmt.Errorf("%w: submission key is required for mutations", ErrInvalidRequest)
	}
	if len(key) > submissionKeyMax {
		return fmt.Errorf("%w: submission key exceeds %d bytes", ErrInvalidRequest, submissionKeyMax)
	}
	for i := 0; i < len(key); i++ {
		if key[i] < 0x20 || key[i] > 0x7e {
			return fmt.Errorf("%w: submission key contains a non-printable byte", ErrInvalidRequest)
		}
	}
	return nil
}

// lookupInput renders the frozen command.get input for a keyed submission:
// {scope, submission_key, operation, operation_version}. The scope is the
// one the original input named, passed through verbatim; every keyed
// operation in the frozen catalog requires one. An input that names no scope
// object cannot be looked up, and the caller keeps the outcome unknown.
func lookupInput(sub submission) (json.RawMessage, error) {
	var original struct {
		Scope json.RawMessage `json:"scope"`
	}
	if err := json.Unmarshal(sub.input, &original); err != nil {
		return nil, fmt.Errorf("command lookup needs the original input's scope: input is not a JSON object")
	}
	if scope := strings.TrimSpace(string(original.Scope)); !strings.HasPrefix(scope, "{") {
		return nil, fmt.Errorf("command lookup needs the original input's scope: input names no scope object")
	}
	return json.Marshal(struct {
		Scope            json.RawMessage `json:"scope"`
		SubmissionKey    string          `json:"submission_key"`
		Operation        string          `json:"operation"`
		OperationVersion int64           `json:"operation_version"`
	}{original.Scope, sub.key, sub.operation, lookupOperationVersion})
}

// retainedCommand mirrors $defs/Command, the command.get output resource.
type retainedCommand struct {
	ID               contract.ID     `json:"id"`
	PrincipalID      contract.ID     `json:"principal_id"`
	Operation        string          `json:"operation"`
	OperationVersion int64           `json:"operation_version"`
	SubmissionKey    string          `json:"submission_key"`
	RequestDigest    contract.Digest `json:"request_digest"`
	Status           string          `json:"status"`
	Data             json.RawMessage `json:"data"`
	ErrorCode        string          `json:"error_code,omitempty"`
	Result           contract.Result `json:"result"`
}

// retainedResult extracts the original disposition from a completed
// command.get envelope. The resource must be the command that was asked for
// (same key, operation and version) and its retained envelope must be
// well formed and carry the command's own identity; anything else proves
// nothing about the original command.
func retainedResult(sub submission, lookup contract.Result) (contract.Result, error) {
	var data struct {
		Resource *retainedCommand `json:"resource"`
	}
	if err := contract.DecodeStrict(lookup.Data, &data); err != nil {
		return contract.Result{}, fmt.Errorf("malformed command resource: %w", err)
	}
	cmd := data.Resource
	if cmd == nil {
		return contract.Result{}, fmt.Errorf("malformed command resource: no resource")
	}
	if cmd.SubmissionKey != sub.key || cmd.Operation != sub.operation || cmd.OperationVersion != lookupOperationVersion {
		return contract.Result{}, fmt.Errorf("command resource names %s v%d, not the submitted command", cmd.Operation, cmd.OperationVersion)
	}
	if err := validateEnvelope(&cmd.Result); err != nil {
		return contract.Result{}, fmt.Errorf("retained %w", err)
	}
	if cmd.Result.CommandID != cmd.ID || cmd.Result.Status != cmd.Status {
		return contract.Result{}, fmt.Errorf("retained result envelope disagrees with its command record")
	}
	return cmd.Result, nil
}

// envelopeStatusByHTTP pins the frozen mapping from HTTP status to the
// result envelope status it may carry: 200 completed, 202 accepted, and the
// documented domain-failure statuses failed. Any other HTTP status is a
// protocol violation; drift against contract.HTTPStatus is covered by the
// fault-matrix test.
var envelopeStatusByHTTP = map[int]string{
	200: contract.StatusCompleted,
	202: contract.StatusAccepted,
	400: contract.StatusFailed,
	403: contract.StatusFailed,
	404: contract.StatusFailed,
	409: contract.StatusFailed,
	410: contract.StatusFailed,
	422: contract.StatusFailed,
	500: contract.StatusFailed,
	503: contract.StatusFailed,
}

// validateResult enforces strict envelope structure and its agreement with
// the HTTP status: the frozen schema string, a durable command identity, an
// exactly-allowed status, error/status consistency, and the HTTP mapping.
func validateResult(httpStatus int, r *contract.Result) error {
	if err := validateEnvelope(r); err != nil {
		return err
	}
	want, ok := envelopeStatusByHTTP[httpStatus]
	if !ok {
		return fmt.Errorf("unexpected HTTP status %d for an operation result", httpStatus)
	}
	if want != r.Status {
		return fmt.Errorf("HTTP status %d disagrees with envelope status %q", httpStatus, r.Status)
	}
	return nil
}

// validateEnvelope enforces the envelope structure that holds with or
// without an HTTP status: the retained envelope inside a command resource
// is checked with it too.
func validateEnvelope(r *contract.Result) error {
	if r.Schema != contract.SchemaResult {
		return fmt.Errorf("result envelope schema %q, want %q", r.Schema, contract.SchemaResult)
	}
	if r.CommandID == "" {
		return fmt.Errorf("result envelope has no command_id")
	}
	switch r.Status {
	case contract.StatusCompleted, contract.StatusAccepted:
		if r.Error != nil {
			return fmt.Errorf("result envelope status %q carries a fault", r.Status)
		}
	case contract.StatusFailed:
		if r.Error == nil {
			return fmt.Errorf("failed result envelope carries no fault")
		}
	default:
		return fmt.Errorf("result envelope status %q is not completed, accepted or failed", r.Status)
	}
	return nil
}

// cursorExpired surfaces a cursor_expired fault as *CursorExpiredError so the
// caller refreshes its snapshot before replaying; every other fault is
// returned unchanged.
func cursorExpired(fault *contract.Fault) error {
	if fault.Code != contract.CodeCursorExpired {
		return fault
	}
	return &CursorExpiredError{Fault: fault, SnapshotRequired: snapshotRequired(fault.Details)}
}

// snapshotRequired reads the open details object for snapshot_required.
// Details are extension space, so the peek is tolerant of other fields and
// of a missing or non-object details value.
func snapshotRequired(details json.RawMessage) bool {
	if len(details) == 0 {
		return false
	}
	var d struct {
		SnapshotRequired bool `json:"snapshot_required"`
	}
	if err := json.Unmarshal(details, &d); err != nil {
		return false
	}
	return d.SnapshotRequired
}

File: apps/desktop/lib/src/transport/envelope.dart
// The versioned request and result envelopes for
// POST /v1/operations/{operation_id}.

import 'dart:convert';

import 'strict_json.dart';

const String requestSchema = 'zatiti.request/v1';
const String resultSchema = 'zatiti.result/v1';

/// The frozen fault vocabulary shared by every transport.
enum FaultCode {
  invalidInput('invalid_input', 400),
  notFound('not_found', 404),
  permissionDenied('permission_denied', 403),
  staleVersion('stale_version', 409),
  reviewRequired('review_required', 409),
  submissionConflict('submission_conflict', 409),
  conflict('conflict', 409),
  cursorExpired('cursor_expired', 410),
  prerequisiteMissing('prerequisite_missing', 422),
  externalActionRequired('external_action_required', 422),
  budgetUnavailable('budget_unavailable', 422),
  capabilityUnsupported('capability_unsupported', 422),
  verificationFailed('verification_failed', 422),
  artifactFault('artifact_fault', 409),
  outcomeUnknown('outcome_unknown', 409),
  controllerUnavailable('controller_unavailable', 503),
  internalError('internal_error', 500),

  /// A code this client build does not know. The wire string stays on
  /// [Fault.wireCode]; the client treats it as a non-retryable failure and
  /// never guesses a meaning for it.
  unrecognized('', 500);

  const FaultCode(this.wire, this.httpStatus);

  /// The frozen string value on the wire.
  final String wire;

  /// The HTTP status the controller renders this fault at.
  final int httpStatus;

  static FaultCode? fromWire(String wire) {
    for (final code in values) {
      if (code != unrecognized && code.wire == wire) return code;
    }
    return null;
  }
}

/// Result payload status values.
enum ResultStatus {
  completed('completed'),
  accepted('accepted'),
  failed('failed');

  const ResultStatus(this.wire);
  final String wire;

  static ResultStatus? fromWire(String wire) {
    for (final s in values) {
      if (s.wire == wire) return s;
    }
    return null;
  }
}

/// A fault carried by a failed result envelope.
class Fault implements Exception {
  const Fault({
    required this.code,
    required this.message,
    required this.retryable,
    this.details,
    String? wireCode,
  }) : _wireCode = wireCode;

  factory Fault.fromJson(Object? json) {
    final o = StrictObject(json, 'fault');
    final wire = o.string('code');
    if (wire.isEmpty) {
      throw StrictJsonException('fault carries an empty code');
    }
    final fault = Fault(
      code: FaultCode.fromWire(wire) ?? FaultCode.unrecognized,
      wireCode: wire,
      message: o.string('message'),
      retryable: o.boolean('retryable'),
      details: o.optionalObject('details'),
    );
    o.finish();
    return fault;
  }

  final FaultCode code;
  final String? _wireCode;
  final String message;
  final bool retryable;

  /// The code string exactly as it arrived.
  String get wireCode => _wireCode ?? code.wire;

  /// Open extension space; inert data, never authority.
  final Map<String, Object?>? details;

  /// The `snapshot_required` detail of a `cursor_expired` fault.
  bool get snapshotRequired => details?['snapshot_required'] == true;

  @override
  String toString() => '$wireCode: $message';
}

/// The result envelope (`zatiti.result/v1`).
class ResultEnvelope {
  const ResultEnvelope({
    required this.commandId,
    required this.status,
    required this.data,
    required this.error,
    required this.nextCursor,
  });

  /// Decodes strictly. Every envelope field is required, no other field is
  /// allowed, and status must agree with the presence of a fault.
  factory ResultEnvelope.fromJson(Object? json) {
    final o = StrictObject(json, 'result envelope');
    final schema = o.string('schema');
    if (schema != resultSchema) {
      throw StrictJsonException(
        'result envelope schema "$schema", want "$resultSchema"',
      );
    }
    final commandId = o.string('command_id');
    if (commandId.isEmpty) {
      throw StrictJsonException('result envelope has no command_id');
    }
    final statusWire = o.string('status');
    final status = ResultStatus.fromWire(statusWire);
    if (status == null) {
      throw StrictJsonException(
        'result envelope status "$statusWire" is not completed, accepted '
        'or failed',
      );
    }
    final rawData = o.nullable('data');
    if (rawData != null && rawData is! Map<String, Object?>) {
      throw StrictJsonException('result envelope data must be object or null');
    }
    final rawError = o.nullable('error');
    final rawCursor = o.nullable('next_cursor');
    if (rawCursor != null && rawCursor is! String) {
      throw StrictJsonException('next_cursor must be string or null');
    }
    o.finish();

    final error = rawError == null ? null : Fault.fromJson(rawError);
    if (status == ResultStatus.failed && error == null) {
      throw StrictJsonException('failed result envelope carries no fault');
    }
    if (status != ResultStatus.failed && error != null) {
      throw StrictJsonException(
        'result envelope status "${status.wire}" carries a fault',
      );
    }
    return ResultEnvelope(
      commandId: commandId,
      status: status,
      data: rawData as Map<String, Object?>?,
      error: error,
      nextCursor: rawCursor as String?,
    );
  }

  final String commandId;
  final ResultStatus status;
  final Map<String, Object?>? data;
  final Fault? error;
  final String? nextCursor;

  /// The data object of a non-failed envelope.
  Map<String, Object?> requireData(String operation) {
    final d = data;
    if (d == null) {
      throw StrictJsonException('$operation returned no data');
    }
    return d;
  }
}

/// The HTTP statuses an operation result may arrive on, and the envelope
/// status each one may carry.
const Map<int, ResultStatus> envelopeStatusByHttp = {
  200: ResultStatus.completed,
  202: ResultStatus.accepted,
  400: ResultStatus.failed,
  403: ResultStatus.failed,
  404: ResultStatus.failed,
  409: ResultStatus.failed,
  410: ResultStatus.failed,
  422: ResultStatus.failed,
  500: ResultStatus.failed,
  503: ResultStatus.failed,
};

/// Checks that [envelope] agrees with the HTTP status it arrived on.
void validateAgainstHttp(int httpStatus, ResultEnvelope envelope) {
  final want = envelopeStatusByHttp[httpStatus];
  if (want == null) {
    throw StrictJsonException(
      'unexpected HTTP status $httpStatus for an operation result',
    );
  }
  if (want != envelope.status) {
    throw StrictJsonException(
      'HTTP status $httpStatus disagrees with envelope status '
      '"${envelope.status.wire}"',
    );
  }
  final fault = envelope.error;
  if (fault != null &&
      fault.code != FaultCode.unrecognized &&
      fault.code.httpStatus != httpStatus) {
    throw StrictJsonException(
      'HTTP status $httpStatus disagrees with fault code '
      '"${fault.wireCode}"',
    );
  }
}

/// Renders the exact request body bytes for one call.
List<int> encodeRequest({
  required Map<String, Object?> input,
  String? submissionKey,
}) {
  return utf8.encode(
    jsonEncode(<String, Object?>{
      'schema': requestSchema,
      if (submissionKey != null) 'submission_key': submissionKey,
      'input': input,
    }),
  );
}

final RegExp _operationIdPattern = RegExp(r'^[a-z0-9_]+(\.[a-z0-9_]+)*$');

/// Accepts the frozen catalog shape only: dot-separated `[a-z0-9_]` segments.
/// The identifier is a path segment, so anything else is refused, not escaped.
bool isValidOperationId(String id) =>
    id.length <= 200 && _operationIdPattern.hasMatch(id);

/// The frozen submission key bound: 1..128 printable ASCII characters.
bool isValidSubmissionKey(String key) {
  if (key.isEmpty || key.length > 128) return false;
  for (final c in key.codeUnits) {
    if (c < 0x20 || c > 0x7E) return false;
  }
  return true;
}
